MoreLinq/ToArrayByIndex.cs

```diff
@@ -126,16 +126,22 @@ static partial class MoreEnumerable
 
             foreach (var e in source)
             {
-                var i = indexSelector(e);
-                if (i < 0)
-                    throw new IndexOutOfRangeException();
-                lastIndex = Math.Max(i, lastIndex);
-                Indexed().Add(new KeyValuePair<int, T>(i, e));
+                switch (indexSelector(e))
+                {
+                    case < 0:
+#pragma warning disable CA2201 // Do not raise reserved exception types
+                        throw new IndexOutOfRangeException();
```

Comment:
`ArgumentOutOfRangeException`

Reply:
This is by-design. It's not an invalid argument of `ToArrayByIndex`, but an invalid index for the target array.

Reply:
In that case, I would recommend intentionally not making the check and allowing the array to throw the exception on the invalid index. The array indexer will repeat the check anyway, and will provide a better exception message than we can provide. Mark a comment that guard-checking is intentionally deferred to the array index instead?

Reply:
The array assignment happens much later and there's not much point in wasting time and resources until then.